include/aos/common/tools/list.hpp

```diff
@@ -0,0 +1,616 @@
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef AOS_LIST_HPP_
+#define AOS_LIST_HPP_
+
+#include <assert.h>
+
+#include "aos/common/tools/buffer.hpp"
+#include "aos/common/tools/error.hpp"
+#include "aos/common/tools/new.hpp"
+
+namespace aos {
+
+template <typename T>
+class List;
+
+/**
+ * List implementation.
+ * @tparam T list type.
+ */
+template <typename T>
+class ListImpl {
+protected:
+    struct Node {
+        bool  mAllocated = false;
+        T     mValue;
+        Node* mNext = nullptr;
+        Node* mPrev = nullptr;
+    };
+
+    /**
+     * List iterator.
+     */
+    class Iterator {
+    public:
+        /**
+         * Creates iterator.
+         */
+        Iterator() = default;
+
+        /**
+         * Creates iterator.
+         *
+         * @param node list node.
+         */
+        explicit Iterator(Node* node)
+            : mCurrentNode(node)
+        {
+        }
+
+        /**
+         * Dereference operator.
+         *
+         * @return T&.
+         */
+        T& operator*() const { return mCurrentNode->mValue; }
+
+        /**
+         * Prefix increment operator.
+         *
+         * @return Iterator&.
+         */
+        Iterator& operator++()
+        {
+            mCurrentNode = mCurrentNode->mNext;
+
+            return *this;
+        }
+
+        /**
+         * Postfix increment operator.
+         *
+         * @return Iterator&.
+         */
+        Iterator operator++(int)
+        {
+            Iterator it(*this);
+
+            operator++();
+
+            return it;
+        }
+
+        /**
+         * Prefix decrement operator.
+         *
+         * @return Iterator&.
+         */
+        Iterator& operator--()
+        {
+            mCurrentNode = mCurrentNode->mPrev;
+
+            return *this;
+        }
+
+        /**
+         * Postfix decrement operator.
+         *
+         * @return Iterator&.
+         */
+        Iterator operator--(int)
+        {
+            Iterator it(*this);
+
+            operator--();
+
+            return it;
+        }
+
+        /**
+         * Checks if iterators are not equal.
+         *
+         * @param other iterator to compare with.
+         * @return bool.
+         */
+        bool operator!=(const Iterator& other) const { return mCurrentNode != other.mCurrentNode; }
+
+    private:
+        friend class List<T>;
+
+        Node* mCurrentNode = nullptr;
+    };
+
+    /**
+     * List const iterator.
+     */
+    class ConstIterator {
+    public:
+        /**
+         *  Creates const iterator.
+         */
+        ConstIterator() = default;
+
+        /**
+         *  Creates const iterator.
+         *
+         * @param node list node.
+         */
+        explicit ConstIterator(const Node* node)
+            : mCurrentNode(const_cast<Node*>(node))
```

Comment:
mCurrentNode can be a pointer to const to avoid const-cast

Reply:
redesigned